NBEP 7: External Memory Management Plugin Interface

gmarkall opened the pull request:
As described in [NBEP 7](https://gmarkall.github.io/numba-nbep7-docs/proposals/external-memory-management.html) and [the EMM documentation](https://gmarkall.github.io/numba-nbep7-docs/cuda/external-memory.html). This implements support for External Memory Management (EMM) Plugins - when an EMM Plugin is in use, Numba does all memory allocation and deallocation through the plugin. The NBEP linked above describes the motivations for doing this in more detail, but the main aim is to enable the use of pool allocators such as those provided by the RAPIDS Memory Manager or CuPy.

The implementation is currently passing all tests with the internal memory management on my system:

```
$ python -m numba.runtests numba.cuda.tests
...
Ran 564 tests in 133.396s

OK (skipped=5)
```

There is a PR to the RAPIDS Memory Manager that implements an EMM Plugin: https://github.com/rapidsai/rmm/pull/317 - the CI is failing on that PR, but this is because the RMM PR tester tests with a release version of Numba (instead of with this branch, which would presently be required). This implementation passes all tests on my system when it is used for memory management:

```
$ NUMBA_CUDA_MEMORY_MANAGER=rmm python -m numba.runtests numba.cuda.tests
...
Ran 564 tests in 142.211s

OK (skipped=11)
```

I have also partially implemented a plugin that enables the use of CuPy allocators: https://github.com/gmarkall/nbep-7/blob/master/nbep7/cupy_mempool.py - this is currently a bit of a "draft", and doesn't support IPC, but demonstrates that the concept functions for more than one memory allocation library. Results with it are:

```
$ NUMBA_CUDA_MEMORY_MANAGER=nbep7.cupy_mempool python -m numba.runtests numba.cuda.tests
...
Ran 564 tests in 111.699s

FAILED (errors=8, skipped=11)
```

The errors are due to the lack of implementation of `get_ipc_handle`. I plan to complete this implementation at some point in the future.

Adding functional tests specifically for the plugin interface is a little bit catch-22 - since testing the interface requires an external library capable of managing CUDA memory, I would expect to add functional tests once an EMM plugin is available in a release version of a memory management library, when it can easily be installed as a dependency for test purposes (possibly to be added when gpuCI testing is up and running, since it will allow CI with CUDA hardware).

For the changes that touched internal memory management (the creation of the `BaseCUDAMemoryManager`, `HostOnlyCUDAMemoryManager`, and `NumbaCUDAMemoryManager` classes, and the related `Context` changes), running of the CUDA test suite in general is exercising these changes.



The NBEP 7 document discusses the code changes and their motivations in general, but a summary of the changes made in this PR, in the order they appear on the diff, for the purposes of aiding reviewers follows:

- Add documentation for the interface and the NBEP 7 document.
- A small fix to the markup in `cuda_array_interface.rst` - I accidentally committed this, but left it in as it fixes a nested list.
- Add a new environment variable, `NUMBA_CUDA_MEMORY_MANAGER`.
- Remove `utils.logger_hasHandlers` - this was a backport from Python 3.5 that is no longer required.
- Change `cuda.defer_cleanup()` so it uses the context's `defer_cleanup()` method, instead of directly disabling deallocations for the context, which would be incorrect when an EMM Plugin is in use.
- Add classes and functions to `driver.py`: `BaseCUDAMemoryManager`, `HostOnlyCUDAMemoryManager`, `NumbaCUDAMemoryManager`, `set_memory_manager`, and `_ensure_memory_manager`. All but the last are described in the NBEP 7 document - the last is called by a context on initialization to set the memory manager class if it has not already been set.
- Make `_SizeNotSet` behave more like an int - the previous implementation defined `__int__` to enable it to be used in arithmetic, but this failed for binary ops with floats, which might be used in EMM plugin implementations.
- Change `_PendingDeallocs` so that it can start out with an unknown capacity, then have the capacity set later. This is because the `HostOnlyCUDAMemoryManager` needs to create an object to hold pending deallocations for the host-side objects, and it may only be later that an EMM plugin subclass can provide the memory capacity.
- Make `_MemoryInfo` public, as it is expected to be instantiated by EMM Plugins.
- The `Context` class now creates an instance of the memory manager (either the internal one or an EMM Plugin depending on settings) and uses it for all allocations and deallocations (`memalloc` etc.) instead of managing memory directly.
- The `Context` no longer has to lazily initialize `self.deallocations`, because it does not need to know the memory capacity for the objects it does manage the lifetime of (events, streams, and modules).
- `defer_cleanup` is added to the `Context` class, which is provided for the purpose of calling into the memory manager's `defer_cleanup` function.
- Various finalizer functions have the `context` variable renamed to `memory_manager`, in order to reflect the type of the object accurately.
- `_StagedIpcImpl.own()` no longer calls `own()` on the memory it returns - this is unnecessary when internal memory management is used, and incorrect when an EMM Plugin is used (discussed in NBEP 7 document).
- Adjust argument list of `__init__` methods for `MemoryPointer`, `MappedMemory`, and `PinnedMemory` to be consistent with each other. 
- Add docstrings in various places throughout `driver.py`, for classes and methods that are part of the public API or referred to in the documentation.
- Add all classes and functions related to EMM Plugins to the public API in the `numba.cuda` module (through `device_init.py`).
- Add a new test skip method to avoid running tests that only apply to internal memory management when an EMM Plugin is in use, and decorate appropriate tests.
- Modify tests that inspected the context to inspect the memory manager instead, where appropriate.

stuartarchibald commented:
@gmarkall sure, Farm ID `numba_smoketest_cuda_30`.

gmarkall commented:
@stuartarchibald A quick ping - did the smoketest finish?

stuartarchibald commented:
@gmarkall apologies for the delay, there was a bit of a stall on the windows builders. Anyway, just checked, the above smoketest passed, 5649ac2 is good.

gmarkall commented:
Many thanks!

gmarkall commented:
Documentation now completed - it can be read with the Sphinx theme at: https://gmarkall.github.io/numba-nbep7-docs/cuda/external-memory.html

stuartarchibald commented:
Thanks for doing the docs. Shall I queue this for review now?

gmarkall commented:
I think it's certainly reviewable now, so I think it would be great to add to the queue rather than waiting longer. I will be making some additions shortly (today), but I don't think the lack of them blocks addition this to the queue. The additions will be:

- [ ] Some additions for testing, but because of the nature of the interface it's tricky to add many unit tests - I think what I do push will mainly be tests that ensure the API doesn't get broken. In one sense the entire test suite checks that the internal memory management (now in `HostOnlyCUDAMemoryManager` and `NumbaCUDAMemoryManager`) is working properly. Functionally testing the plugin interface requires a whole external library that can manage CUDA memory, so this may be something that will have to come later (perhaps when gpuCI testing is up and working).
- [ ] I'll add the text of the NBEP to the list of NBEPs section of the documentation - whilst the regular documentation contains a lot of information that's in the NBEP, the NBEP goes into more detail about the rationale for various things, so I imagine it's useful to have in there.
- [ ] I'll also add a comment with a summary of the changes made, in order to aid review.

gmarkall commented:
I think this is now ready for review so I've updated the summary and removed [WIP] from the subject. I haven't added any specific tests for the memory management (see discussion in the summary about the catch-22 nature of adding functional tests for the EMM Plugin interface) but would be keen to add a functional test in a subsequent PR as soon as is practical.

leofang commented on docs/source/cuda/external-memory.rst:
```diff
@@ -0,0 +1,280 @@
+======================================
+
+When an EMM Plugin is in use (see :ref:`setting-emm-plugin`), Numba will make
+does allocation and deallocation through the Plugin. It will never directly call
```
```suggestion
allocation and deallocation through the Plugin. It will never directly call
```

gmarkall commented:
I've added some tests but they're failing on CI at present because some of the definitions in the test module need guarding to prevent them being used on the simulator - will fix this up tomorrow.

stuartarchibald commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -1348,7 +1645,25 @@ def own(self):
+    :type finalizer: function
+    """
+
+    def __init__(self, context, pointer, size, owner=None, finalizer=None):
```
Is this an API change, any idea of prevalence of use elsewhere, does it need a deprecation cycle?

gmarkall replied:
It was an internal API before this PR, so my inclination is not to give it a deprecation cycle.

stuartarchibald replied:
agree

stuartarchibald commented on numba/cuda/testing.py:
```diff
@@ -30,6 +30,10 @@ def skip_unless_conda_cudatoolkit(reason):
     return unittest.skipUnless(get_conda_ctk() is not None, reason)
 
 
+def skip_with_external_memmgr(reason):
```
Docstring? Also, should it be called `skip_if_external_memmgr` or `skip_on_external_memmgr` maybe?

gmarkall replied:
There's no docstring for any skip method in `numba.cuda.testing`, but I can add them. I'll go with `skip_if_external_memmgr`, which I think sounds better.

gmarkall replied:
Renamed to `skip_if_external_memmgr` and added docstrings.

stuartarchibald commented on numba/cuda/tests/cudadrv/test_emm_plugins.py:
```diff
@@ -0,0 +1,188 @@
+from numba import cuda
+from numba.cuda.testing import unittest, SerialMixin, skip_on_cudasim
+
+not_linux = not sys.platform.startswith('linux')
```
Suggest removing this and altering the test that makes use of it to use `@numba.tests.support.linux_only`.

gmarkall replied:
Will do - I copied this from `test_ipc` so I'll fix it there too.

gmarkall replied:
I just realised that `linux_only` is the opposite of `not_linux`, so I'll add `not_linux` to `numba.tests.support` instead.

gmarkall replied:
I've got confused by the double negative of skipping on not linux. Never mind!

gmarkall replied:
Now uses `@linux_only`.

stuartarchibald commented:
Once public CI passes, please ping me and I'll run it through the build farm, thanks!

gmarkall commented on docs/source/cuda/external-memory.rst:
```diff
@@ -0,0 +1,280 @@
+
+When an EMM Plugin is in use (see :ref:`setting-emm-plugin`), Numba will make
+allocation and deallocation through the Plugin. It will never directly call
+functions such as ``cuMemAlloc``, ``cuMemFree``, etc.
```
Just to clarify this scenario, is it:

- An extension library uses `numba.cuda` to call `cuMemAlloc` to allocate memory, and then later calls `cuMemFree` on its own allocations,
- An extension library uses `numba.cuda` to call `cuMemFree` on memory pointers provided by the EMM Plugin, or
- Something else?

In any case, the first scenario is fine - behaviour will be well defined, but if a third-party library doesn't use the same allocator that Numba does, there's not a lot that can be done to stop it. The second is obviously wrong, since the pointers might well be part of a pool, and the EMM plugin will be expecting to free them at some point, but I'm finding it hard picture a use case where you would expect to do this.

stuartarchibald replied:
I think this is ok on second thoughts. Re-read it as "Numba internally will make allocations through the EMM" and that's fine, further nothing precludes a users using `numba.cuda.cuMemAlloc/cuMemFree` independently.

gmarkall commented on docs/source/cuda/external-memory.rst:
```diff
@@ -0,0 +1,280 @@
+
+In addition to memory, Numba manages the allocation and deallocation of
+:ref:`events <events>`, :ref:`streams <streams>`, and modules (modules are a
+compiled objects, which is generated for CUDA kernels). The management of
```
I completely agree :-)

Will fix.

gmarkall commented on docs/source/cuda/external-memory.rst:
```diff
@@ -0,0 +1,280 @@
+
+The Host-Only CUDA Memory Manager
+---------------------------------
+
```
The memory manager is never host-only - that class is just to provide something to base an implementation of an EMM Plugin for libraries that don't want to manage host memory - e.g. the internal `NumbaCUDAMemoryManager` subclasses it, as does the RMM EMM Plugin (because RMM only does device memory at present).

There isn't a way to query what the memory manager is at the moment - ideally it should be transparent to the Numba user, so I'm finding it hard to see why you'd do something different when different memory managers are in use. I could add a `__repr__`, but if a `__repr__` should be provided, it's probably better to recommend that EMM Plugin authors implement `__repr__`, as `HostOnlyCUDAMemoryManager` is never instantiated directly (if you try the abstract base class checking will complain that `memalloc` etc. are not implemented).

stuartarchibald replied:
Perhaps that was a bad example, it was more about "is there likely to be a useful control flow pattern that is based on the type of memory manager", I now realise that this is unlikely the case and most probably something authors can just implement as they see fit.

gmarkall commented on docs/source/cuda/external-memory.rst:
```diff
@@ -0,0 +1,280 @@
+Plugin:
+
+- The ``memory`` parameter to ``get_ipc_handle`` can be passed as the ``base``
+  parameter.
```
The `get_ipc_handle` method of an EMM plugin receives a parameter called `memory`, which you can pass as the `IpcHandle` constructor's `base` parameter. Will edit the text to make this clearer.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+.. _nbep-7:
+
+===========================================
+NBEP 7: External Memory Management Plugins
```
Indeed it should!

gmarkall commented on docs/source/cuda/external-memory.rst:
```diff
@@ -0,0 +1,280 @@
+It is recommended that the memory manager is set once prior to using any CUDA
+functionality, and left unchanged for the remainder of execution.
+
+.. autofunction:: numba.cuda.set_memory_manager
```
You're welcome - thanks for a thorough reading!

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+* Changing the memory manager once execution has begun. It is not practical to
+  change the memory manager and retain all allocations. Cleaning up the entire
+  state and then changing to a different memory allocator (rather than starting
+  a new process) appears to be a rather niche use case.
```
It's possible that that could happen. Originally I thought about making the choice of memory manager fixed once it is set once, but it's useful for testing to be able to change it. I thought about adding a warning if the memory manager is changed, but I'd expect this to result in warnings from the test suite.

As it's unlikely to be a common issue, I'm inclined not to add anything to check for this immediately - if it becomes a problem in use, then I'd think about examining the use cases where it is a problem to decide what an appropriate check looks like - would you be happy with this approach?

stuartarchibald replied:
I agree, a later version can work on how to handle this if required, just wanted to ensure it was covered as a possibility.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+descriptions above and to illustrate how the plugin interface can be used -
+different choices may be made for a production-ready implementation.
+
+The plugin implementation consists of additions to ``python/rmm/rmm.py``\ :
```
It is - I've added one.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+       def memalloc(self, size):
+           # Allocates device memory using RMM functions. The finalizer for the
+           # allocated memory calls back to RMM to free the memory.
+           addr = librmm.rmm_alloc(bytesize, 0)
```
`bytesize` should have been `size` - will fix. `librmm` is already imported in `rmm.py`, so I didn't put the import in the example as it's just additions to `rmm.py`.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+
+   # New imports:
+   from contextlib import context_manager
+   from numba.cuda import HostOnlyCUDAMemoryManager, MemoryPointer, IpcHandle
```
Numba is already a dependency of RMM - I'll add a note of this.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+
+       def get_memory_info(self):
+           # Returns a tuple of (free, total) using RMM functionality.
+           return get_memory_info()
```
Now I look at it, that should have actually been `get_info`, which is a function from rmm.py - I've changed the example and added a note that it's defined in rmm.py.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+
+   # Utility function register `RMMNumbaManager` as an EMM:
+   def use_rmm_for_numba():
+       cuda.set_memory_manager(RMMNumbaManager)
```
The one already imported by rmm.py. For clarity though, I've just changed it so that `set_memory_manager` is explicitly imported so it can be seen as a new addition.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+:class:`~numba.cuda.cudadrv.driver.Context` class. It maintains lists of
+allocations and deallocations:
+
+* ``allocations`` is a ``numba.utils.UniqueDict``, created at context creation time.
```
Renamed.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+
+These are used to track allocations and deallocations of:
+
+
```
Done.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+This method arguably un-needed - pinned memory is immediately unpinned by its
+finalizer, and unpinning before a finalizer runs would invalidate the state of
+``PinnedMemory`` objects for which references are still held. It is proposed that
+this is removed when making the other changes to the ``Context`` class.
```
Thanks for confirming.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+
+* Numba branch: https://github.com/gmarkall/numba/tree/grm-numba-nbep-7.
+* RMM branch: https://github.com/gmarkall/rmm/tree/grm-numba-nbep-7.
+* CuPy implementation: `nbep7/cupy_mempool.py <nbep7/cupy_mempool.py>`_ - uses an
```
Ah, I used an MD to RST translator on the original proposal, and since that was part of the same repo the link ended up being relative - will fix!

gmarkall replied:
Fixed now.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+
+.. code-block::
+
+   NUMBA_CUDA_MEMORY_MANAGER=nbep7. python -m numba.runtests numba.cuda.tests
```
It's `cupy_mempool` - now fixed.

gmarkall commented on numba/core/config.py:
```diff
@@ -357,6 +357,9 @@ def avx_default():
         GDB_BINARY = _readenv("NUMBA_GDB_BINARY", str, '/usr/bin/gdb')
 
+        # CUDA Memory management
+        CUDA_MEMORY_MANAGER = _readenv("NUMBA_CUDA_MEMORY_MANAGER", str, '')
```
I think that's a good idea - will make that the case (and modify the check in `numba.cuda.cudadrv.driver`.

gmarkall commented on numba/core/utils.py:
```diff
@@ -421,21 +421,6 @@ def benchmark(func, maxsec=1):
 
 
-def logger_hasHandlers(logger):
-    # Backport from python3.5 logging implementation of `.hasHandlers()`
```
:-)

There's a couple of other backports and workarounds in the CUDA target for earlier versions that I've come across, but have yet to remove - will feature these in future PRs :-)

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+        ready to use.
+
+        :return: None
+        """
```
I think that role is taken by the `reset` method - it cleans up all allocations in the context, which is everything that an EMM Plugin instance is responsible for.

stuartarchibald replied:
ok, thanks, let's leave.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+        """
+        Returns an integer specifying the version of the EMM Plugin interface
+        supported by the plugin implementation. Should always return 1 for
+        implementations described in this proposal.
```
Yes - will fix.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+            else:
+                raise
+
+    def memhostalloc(self, size, mapped=False, portable=False,
```
Thanks for confirming.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+        """
+        if isinstance(pointer, (int, long)):
+            pointer = c_void_p(pointer)
+
```
I intended to keep this in the `Context.mempin` method but have accidentally removed it during all my changes - will put it back. Great catch!

gmarkall replied:
I've now put this back into `Context.mempin`.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+
+    @contextlib.contextmanager
+    def defer_cleanup(self):
+        """Disables cleanup of mapped or pinned host memory in the current
```
It should be "Returns a context manager that disables cleanup of mapped or pinned host memory in the current context whilst it is active." - will fix.

gmarkall replied:
Docstring now updated.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+    def initialize(self):
+        # Set the memory capacity of *deallocations* as the memory manager
+        # becomes active for the first time
+        if self.deallocations.memory_capacity == _SizeNotSet:
```
Now that `_SizeNotSet` is more int-like than before, I felt that `==` was more appropriate. This could get confused with a device that has a memory capacity of 0, but such a device doesn't exist. I'm happy to change it back to `is` if you'd prefer?

stuartarchibald replied:
I think your reasoning is fine. TBH, I'm more concerned about the module globals having a potential race condition on the type of `_SizeNotSet`!

gmarkall replied:
I'm not sure what you're referring to regarding a potential race condition?

stuartarchibald replied:
Oh that there's two definitions for the same global and its protected by virtue of the import lock impl. Not of concern for this PR.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+        if self.deallocations.memory_capacity == _SizeNotSet:
+            self.deallocations.memory_capacity = self.get_memory_info().total
+
+    def memalloc(self, size):
```
Many thanks.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+        driver.cuMemGetInfo(byref(free), byref(total))
+        return MemoryInfo(free=free.value, total=total.value)
+
+    def get_ipc_handle(self, memory):
```
The `Context` class already checked whether it did: https://github.com/numba/numba/pull/5328/files#diff-932d549c42cb64d05366984424e13dcfL841

stuartarchibald replied:
This link doesn't appear to point to anything?!

gmarkall replied:
Maybe this does: https://github.com/numba/numba/pull/5328/files#diff-932d549c42cb64d05366984424e13dcfR1098 - line 841 / 1098 of driver.py - 

```python
    def get_ipc_handle(self, memory):
        """
        Returns a *IpcHandle* from a GPU allocation.
        """
        if not SUPPORTS_IPC:
            raise OSError('OS does not support CUDA IPC')
        return self.memory_manager.get_ipc_handle(memory)
```

The check was done right before `get_ipc_handle` was called.

stuartarchibald replied:
Thanks, got it.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+    dummy = mm_plugin(context=None)
+    iv = dummy.interface_version
+    if iv != 1:
+        err = "EMM Plugin interface has version %d - version 1 required" % iv
```
Good point.

gmarkall replied:
Done.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+        return 1
+
+
+_memory_manager = None
```
Will change this and update all references.

gmarkall replied:
Having re-thought this - `_MEMORY_MANAGER` implies a constant to me, but it is not constant (it could be set first). Do you still want it uppercasing given that it can change?

stuartarchibald replied:
ah ok, yes, on second thoughts, that it's non-const suggests lower case is fine, let's leave.

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+            raise RuntimeError("Failed to use memory manager from %s" %
+                               config.CUDA_MEMORY_MANAGER)
+    else:
+        _memory_manager = NumbaCUDAMemoryManager
```
It is, but that would be a pretty weird thing to be doing. I would suggest it doesn't matter.

stuartarchibald replied:
ok, am happy to take your opinion, just so long as we've considered it :)

gmarkall commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -1232,18 +1488,39 @@ def _rebuild(cls, handle_ary, size, source_info, offset):
     __cuda_memory__ = True
 
-    def __init__(self, context, pointer, size, finalizer=None, owner=None):
+    def __init__(self, context, pointer, size, owner=None, finalizer=None):
```
It makes the order of arguments is now consistent across `MemoryPointer`, `AutoFreePointer`, `MappedMemory`, and `PinnedMemory` - since this is bringing a private API into the public it seemed like a good time to do this cleanup (rather than leaving an inconsistent API).

stuartarchibald replied:
I'm happy with this reasoning, and agree with proposed change.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+
+   Event Type,Device ID,Address,Stream,Size (bytes),Free Memory,Total Memory,Current Allocs,Start,End,Elapsed,Location
+   Alloc,0,0x7fae06600000,0,80,0,0,1,1.10549,1.1074,0.00191666,/home/nfs/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py:683
+   Free,0,0x7fae06600000,0,0,0,0,0,1.10798,1.10921,0.00122238,/home/nfs/gmarkall/numbadev/numba/numba/utils.py:678
```
Is there a reason to do this? I was trying to give a representative example of output from running with the external memory manager - for the documentation, I can imagine paths might be a little less desirable, but for the NBEP document I'm not sure I understand why it would matter.

stuartarchibald replied:
Largely just consistency with elsewhere. `egrep -r 'path>' docs/source/`.

gmarkall replied:
Paths now anonymised, replaced with `<path>`.

gmarkall commented on docs/source/proposals/external-memory-management.rst:
```diff
@@ -0,0 +1,918 @@
+
+   Event Type,Device ID,Address,Stream,Size (bytes),Free Memory,Total Memory,Current Allocs,Start,End,Elapsed,Location
+   Alloc,0,0x7f96c7400000,0,80,0,0,1,1.13396,1.13576,0.00180059,/home/nfs/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py:686
+   Free,0,0x7f96c7400000,0,0,0,0,0,1.13628,1.13723,0.000956004,/home/nfs/gmarkall/numbadev/numba/numba/utils.py:678
```
Same answer :-)

gmarkall replied:
Now replaced.

gmarkall commented:
@stuartarchibald Thanks for the thorough review! The tests are now passing on CI, and I've addressed most of the comments - the ones that are addressed straightforwardly I've marked as resolved. Some comments are outstanding, and I've left those unresolved at present.

I think this is a reasonable time for a buildfarm run, as it doesn't look like there's much left to change.

stuartarchibald commented:
Buildfarm ID: `numba_smoketest_cuda_38`.

stuartarchibald commented:
62e6a07 is failing on the farm. Only Linux systems fail. Failing test:
`numba.cuda.tests.cudapy.test_ipc.TestIpcNotSupported`
Fail reason
`AssertionError: OSError not raised`.

gmarkall commented:
Thanks for the test! Just pushed a fix and waiting for the normal CI.

stuartarchibald commented on numba/cuda/tests/cudapy/test_ipc.py:
```diff
@@ -180,7 +180,7 @@ def test_ipc_array(self):
         self.check_ipc_array(slice(None, 8))
 
-@linux_only
+@unittest.skipUnless(not_linux, 'Only on OS other than Linux')
```
For double negative avoidance, would `@unittest.skipIf(linux)` be ok (where `linux` is defined like above, but without the `not`!?

gmarkall replied:
Yes, I got confused around this once already!

gmarkall replied:
Fix pushed, waiting for CI again.

stuartarchibald commented:
c406168 is Buildfarm ID: `numba_smoketest_cuda_39`.

stuartarchibald commented:
Buildfarm passed on https://github.com/numba/numba/commit/c40616842de024cd6db59b5ab9fd14abe9b42e05.

stuartarchibald commented:
@sklam This seems good to me, buildfarm is also happy, will leave it with you, thanks!

stuartarchibald commented:
Thanks all, marking for merge.

sklam commented on numba/cuda/cudadrv/driver.py:
```diff
@@ -633,8 +976,17 @@ def __len__(self):
 
-_MemoryInfo = namedtuple("_MemoryInfo", "free,total")
+MemoryInfo = namedtuple("MemoryInfo", "free,total")
+"""Free and total memory for a device.
```
I have never realized that this is the convention for docstring on global variables. Sphinx is picking it up.

gmarkall replied:
It's the first time I've used it - I found it when I was searching for how to document a `namedtuple`.
